### PR TITLE
Fix type error hidden by implicit downcasts

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

There is a type error hidden by an implicit downcast in the following line (this was caught via crash reporting):

```
int.parse(sdkMatch?.group(2) ?? 11)
```

The least upper bound of `String` (returned from the regex match) and `int` is Object, which then gets implicitly downcast back to String for int.parse. This is incorrect though, as the 11 should be quoted or tryParse used instead.

Cleanup: make file implicit downcast free.

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 1]
```diff
@@ -463,14 +463,14 @@ class IOSSimulator extends Device {
 
   Future<int> get sdkMajorVersion async {
     final Match sdkMatch = _iosSdkRegExp.firstMatch(await sdkNameAndVersion);
-    return int.parse(sdkMatch?.group(2) ?? 11);
+    return int.parse(sdkMatch?.group(2) ?? '11');
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: I had this before but tryParse can't handle null values. Based on the regex, it seems unlikely that group(2) would ever be non-null and contain an invalid number. The correct code with all branches covered would be something like:

```
final String group = sdkMatch?.group(2);
if (group == null) {
  return 11;
}
return int.tryParse(group) ?? 11;
```

But it just felt so verbose

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 2]
```diff
@@ -269,7 +269,7 @@ class IOSSimulator extends Device {
   @override
   Future<bool> installApp(ApplicationPackage app) async {
     try {
-      final IOSApp iosApp = app;
+      final IOSApp iosApp = app as IOSApp;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 3]
```diff
@@ -371,7 +371,7 @@ class IOSSimulator extends Device {
       // which should always yield the correct value and does not require
       // parsing the xcodeproj or configuration files.
       // See https://github.com/flutter/flutter/issues/31037 for more information.
-      final IOSApp iosApp = package;
+      final IOSApp iosApp = package as IOSApp;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 4]
```diff
@@ -412,7 +412,7 @@ class IOSSimulator extends Device {
         extraGenSnapshotOptions: buildInfo.extraGenSnapshotOptions);
 
     final XcodeBuildResult buildResult = await buildXcodeProject(
-      app: app,
+      app: app as BuildableIOSApp,
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 5]
```diff
@@ -422,7 +422,7 @@ class IOSSimulator extends Device {
       throwToolExit('Could not build the application for the simulator.');
 
     // Step 2: Assert that the Xcode project was successfully built.
-    final IOSApp iosApp = app;
+    final IOSApp iosApp = app as IOSApp;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 6]
```diff
@@ -463,14 +463,14 @@ class IOSSimulator extends Device {
...
-    return int.parse(sdkMatch?.group(2) ?? 11);
+    return int.parse(sdkMatch?.group(2) ?? '11');
   }
 
   @override
   DeviceLogReader getLogReader({ ApplicationPackage app }) {
     assert(app is IOSApp);
     _logReaders ??= <ApplicationPackage, _IOSSimulatorLogReader>{};
-    return _logReaders.putIfAbsent(app, () => _IOSSimulatorLogReader(this, app));
+    return _logReaders.putIfAbsent(app, () => _IOSSimulatorLogReader(this, app as IOSApp));
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 7]
```diff
@@ -107,7 +107,7 @@ class SimControl {
     final Map<String, dynamic> devicesSection = await _list(SimControlListSection.devices);
 
     for (String deviceCategory in devicesSection.keys) {
-      final List<dynamic> devicesData = devicesSection[deviceCategory];
+      final List<dynamic> devicesData = devicesSection[deviceCategory] as List<dynamic>;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 8]
```diff
@@ -90,7 +90,7 @@ class SimControl {
       return <String, Map<String, dynamic>>{};
     }
     try {
-      return json.decode(results.stdout)[section.name];
+      return json.decode(results.stdout?.toString())[section.name] as Map<String, dynamic>;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 9]
```diff
@@ -107,9 +112,11 @@ class SimControl {
     final Map<String, dynamic> devicesSection = await _list(SimControlListSection.devices);
 
     for (String deviceCategory in devicesSection.keys) {
-      final List<dynamic> devicesData = devicesSection[deviceCategory] as List<dynamic>;
-      for (Map<String, dynamic> data in devicesData.map<Map<String, dynamic>>(castStringKeyedMap)) {
-        devices.add(SimDevice(deviceCategory, data));
+      final Object devicesData = devicesSection[deviceCategory];
+      if (devicesData != null && devicesData is List<dynamic>) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: nit: Flip the sense of the if and unindent the nested for loop:
```dart
if (devicesData == null || devicesData is! List<dynamic>) {
  continue;
}
...
```

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 10]
```diff
@@ -325,7 +332,7 @@ class IOSSimulator extends Device {
     bool usesTerminalUi = true,
     bool ipv6 = false,
   }) async {
-    if (!prebuiltApplication) {
+    if (!prebuiltApplication && package is BuildableIOSApp) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: If it's not prebuilt and not buildable, then how did we get here? Should the 'is' check be an assert?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: we need the `is` check to make this not an implicit cast, even with the covariant on _setupUpdatedApplicationBundle

File: packages/flutter_tools/lib/src/ios/simulators.dart (hunk 9 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: discussed offline: dart will not type promote outside of the block to `List<dynamic>`, thus requiring another cast

**Pull request merged**